File: packages/java/dotdo/src/main/java/com/dotdo/DotDo.java
package com.dotdo;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.time.Instant;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Function;
import java.util.function.Supplier;

/**
 * DotDo Platform SDK for Java 21+.
 *
 * <p>The main entry point for interacting with the DotDo platform.
 * Provides authentication, connection pooling, retry logic, and high-level APIs.
 *
 * <p>Example usage:
 * <pre>{@code
 * // Create and configure the client
 * var client = DotDo.builder()
 *     .apiKey("your-api-key")
 *     .baseUrl("https://api.dotdo.com")
 *     .build();
 *
 * // Make authenticated requests
 * var result = client.call("todos.list").await();
 *
 * // With automatic retry
 * var created = client.withRetry(3)
 *     .call("todos.create", Map.of("title", "Buy milk"))
 *     .await();
 * }</pre>
 */
public final class DotDo implements AutoCloseable {

    private final Config config;
    private final HttpClient httpClient;
    private final ConnectionPool connectionPool;
    private final AuthManager authManager;
    private final RetryPolicy retryPolicy;
    private final ScheduledExecutorService scheduler;

    private volatile boolean closed = false;

    private DotDo(Config config) {
        this.config = config;
        this.httpClient = HttpClient.newBuilder()
                .connectTimeout(config.connectTimeout())
                .executor(config.virtualThreads()
                        ? Executors.newVirtualThreadPerTaskExecutor()
                        : Executors.newCachedThreadPool())
                .build();
        this.connectionPool = new ConnectionPool(config.maxConnections());
        this.authManager = new AuthManager(config);
        this.retryPolicy = new RetryPolicy(config.maxRetries(), config.retryDelay());
        this.scheduler = Executors.newSingleThreadScheduledExecutor();
    }

    /**
     * Create a new DotDo client builder.
     *
     * @return a new builder instance
     */
    public static Builder builder() {
        return new Builder();
    }

    /**
     * Create a client with default configuration and API key.
     *
     * @param apiKey the API key for authentication
     * @return a new DotDo client
     */
    public static DotDo create(String apiKey) {
        return builder().apiKey(apiKey).build();
    }

    // =========================================================================
    // API Methods
    // =========================================================================

    /**
     * Make an authenticated RPC call.
     *
     * @param method the method name (e.g., "todos.list", "users.get")
     * @return a promise for the response
     */
    public Promise<Response> call(String method) {
        return call(method, null);
    }

    /**
     * Make an authenticated RPC call with parameters.
     *
     * @param method the method name
     * @param params the method parameters
     * @return a promise for the response
     */
    public Promise<Response> call(String method, Object params) {
        ensureNotClosed();

        return new Promise<>(retryPolicy.execute(() -> {
            var connection = connectionPool.acquire();
            try {
                String token = authManager.getToken();
                String jsonBody = buildJsonRpcRequest(method, params);

                var request = HttpRequest.newBuilder()
                        .uri(URI.create(config.baseUrl() + "/rpc"))
                        .header("Content-Type", "application/json")
                        .header("Accept", "application/json")
                        .header("Authorization", "Bearer " + token)
                        .POST(HttpRequest.BodyPublishers.ofString(jsonBody))
                        .timeout(config.requestTimeout())
                        .build();

                return httpClient.sendAsync(request, HttpResponse.BodyHandlers.ofString())
                        .thenApply(response -> {
                            if (response.statusCode() == 401) {
                                authManager.invalidateToken();
                                throw new AuthException("Authentication failed");
                            }
                            if (response.statusCode() >= 200 && response.statusCode() < 300) {
                                return parseResponse(response.body());
                            }
                            throw new DotDoException("HTTP " + response.statusCode() + ": " + response.body());
                        });
            } finally {
                connectionPool.release(connection);
            }
        }));
    }

    /**
     * Create a client with custom retry behavior.
     *
     * @param maxRetries maximum number of retry attempts
     * @return a client wrapper with custom retry settings
     */
    public RetryClient withRetry(int maxRetries) {
        return new RetryClient(this, maxRetries);
    }

    /**
     * Create a client with custom timeout.
     *
     * @param timeout the request timeout
     * @return a client wrapper with custom timeout
     */
    public TimeoutClient withTimeout(Duration timeout) {
        return new TimeoutClient(this, timeout);
    }

    // =========================================================================
    // Resource Management
    // =========================================================================

    private void ensureNotClosed() {
        if (closed) {
            throw new IllegalStateException("DotDo client has been closed");
        }
    }

    @Override
    public void close() {
        closed = true;
        scheduler.shutdown();
        connectionPool.close();
        try {
            if (!scheduler.awaitTermination(5, TimeUnit.SECONDS)) {
                scheduler.shutdownNow();
            }
        } catch (InterruptedException e) {
            scheduler.shutdownNow();
            Thread.currentThread().interrupt();
        }
    }

    // =========================================================================
    // JSON Serialization
    // =========================================================================

    private static final AtomicInteger requestIdGenerator = new AtomicInteger(0);

    private String buildJsonRpcRequest(String method, Object params) {
        int id = requestIdGenerator.incrementAndGet();
        var sb = new StringBuilder();
        sb.append("{\"jsonrpc\":\"2.0\",\"id\":").append(id);
        sb.append(",\"method\":\"").append(escapeJson(method)).append("\"");
        if (params != null) {
            sb.append(",\"params\":").append(serializeParams(params));
        }
        sb.append("}");
        return sb.toString();
    }

    private String serializeParams(Object params) {
        if (params == null) return "null";
        if (params instanceof String s) return "\"" + escapeJson(s) + "\"";
        if (params instanceof Number) return params.toString();
        if (params instanceof Boolean) return params.toString();
        if (params instanceof Map<?, ?> map) {
            var sb = new StringBuilder("{");
            boolean first = true;
            for (var entry : map.entrySet()) {
                if (!first) sb.append(",");
                first = false;
                sb.append("\"").append(escapeJson(entry.getKey().toString())).append("\":");
                sb.append(serializeParams(entry.getValue()));
            }
            sb.append("}");
            return sb.toString();
        }
        if (params instanceof Iterable<?> list) {
            var sb = new StringBuilder("[");
            boolean first = true;
            for (var item : list) {
                if (!first) sb.append(",");
                first = false;
                sb.append(serializeParams(item));
            }
            sb.append("]");
            return sb.toString();
        }
        return "null";
    }

    private String escapeJson(String s) {
        return s.replace("\\", "\\\\")
                .replace("\"", "\\\"")
                .replace("\n", "\\n")
                .replace("\r", "\\r")
                .replace("\t", "\\t");
    }

    private Response parseResponse(String json) {
        // Simplified parsing - use Gson in production
        Object result = null;
        Object error = null;

        if (json.contains("\"error\":")) {
            int errorIdx = json.indexOf("\"error\":");
            if (errorIdx >= 0 && !json.substring(errorIdx).startsWith("\"error\":null")) {
                error = extractJsonValue(json, errorIdx + 8);
            }
        }

        if (json.contains("\"result\":")) {
            int resultIdx = json.indexOf("\"result\":");
            if (resultIdx >= 0) {
                result = extractJsonValue(json, resultIdx + 9);
            }
        }

        return new Response(result, error);
    }

    private Object extractJsonValue(String json, int start) {
        while (start < json.length() && Character.isWhitespace(json.charAt(start))) {
            start++;
        }
        if (start >= json.length()) return null;

        char c = json.charAt(start);
        if (c == '"') {
            int end = start + 1;
            while (end < json.length() && json.charAt(end) != '"') {
                if (json.charAt(end) == '\\') end++;
                end++;
            }
            return json.substring(start + 1, end);
        } else if (c == '{' || c == '[') {
            int depth = 1;
            int end = start + 1;
            while (end < json.length() && depth > 0) {
                char ch = json.charAt(end);
                if (ch == '{' || ch == '[') depth++;
                else if (ch == '}' || ch == ']') depth--;
                else if (ch == '"') {
                    end++;
                    while (end < json.length() && json.charAt(end) != '"') {
                        if (json.charAt(end) == '\\') end++;
                        end++;
                    }
                }
                end++;
            }
            return json.substring(start, end);
        } else if (c == 'n') {
            return null;
        } else if (c == 't') {
            return true;
        } else if (c == 'f') {
            return false;
        } else {
            int end = start;
            while (end < json.length() && (Character.isDigit(json.charAt(end)) || json.charAt(end) == '.' || json.charAt(end) == '-')) {
                end++;
            }
            String numStr = json.substring(start, end);
            if (numStr.contains(".")) {
                return Double.parseDouble(numStr);
            } else {
                return Long.parseLong(numStr);
            }
        }
    }

    // =========================================================================
    // Configuration
    // =========================================================================

    /**
     * Client configuration.
     */
    public record Config(
            String baseUrl,
            String apiKey,
            String clientId,
            String clientSecret,
            Duration connectTimeout,
            Duration requestTimeout,
            int maxConnections,
            int maxRetries,
            Duration retryDelay,
            boolean virtualThreads
    ) {
        public static Config defaults() {
            return new Config(
                    "https://api.dotdo.com",
                    null,
                    null,
                    null,
                    Duration.ofSeconds(10),
                    Duration.ofSeconds(30),
                    10,
                    3,
                    Duration.ofMillis(100),
                    true
            );
        }
    }

    /**
     * Builder for DotDo client.
     */
    public static final class Builder {
        private String baseUrl = "https://api.dotdo.com";
        private String apiKey = null;
        private String clientId = null;
        private String clientSecret = null;
        private Duration connectTimeout = Duration.ofSeconds(10);
        private Duration requestTimeout = Duration.ofSeconds(30);
        private int maxConnections = 10;
        private int maxRetries = 3;
        private Duration retryDelay = Duration.ofMillis(100);
        private boolean virtualThreads = true;

        /**
         * Set the base URL for the DotDo API.
         */
        public Builder baseUrl(String url) {
            this.baseUrl = url.endsWith("/") ? url.substring(0, url.length() - 1) : url;
            return this;
        }

        /**
         * Set the API key for authentication.
         */
        public Builder apiKey(String apiKey) {
            this.apiKey = apiKey;
            return this;
        }

        /**
         * Set OAuth2 client credentials.
         */
        public Builder oauth(String clientId, String clientSecret) {
            this.clientId = clientId;
            this.clientSecret = clientSecret;
            return this;
        }

        /**
         * Set the connection timeout.
         */
        public Builder connectTimeout(Duration timeout) {
            this.connectTimeout = timeout;
            return this;
        }

        /**
         * Set the request timeout.
         */
        public Builder requestTimeout(Duration timeout) {
            this.requestTimeout = timeout;
            return this;
        }

        /**
         * Set the maximum number of connections in the pool.
         */
        public Builder maxConnections(int max) {
            this.maxConnections = max;
            return this;
        }

        /**
         * Set the maximum number of retries.
         */
        public Builder maxRetries(int max) {
            this.maxRetries = max;
            return this;
        }

        /**
         * Set the delay between retries.
         */
        public Builder retryDelay(Duration delay) {
            this.retryDelay = delay;
            return this;
        }

        /**
         * Enable or disable virtual threads (default: enabled).
         */
        public Builder virtualThreads(boolean enabled) {
            this.virtualThreads = enabled;
            return this;
        }

        /**
         * Build the DotDo client.
         */
        public DotDo build() {
            if (apiKey == null && clientId == null) {
                throw new IllegalStateException("Either apiKey or oauth credentials must be provided");
            }
            var config = new Config(
                    baseUrl, apiKey, clientId, clientSecret,
                    connectTimeout, requestTimeout, maxConnections,
                    maxRetries, retryDelay, virtualThreads
            );
            return new DotDo(config);
        }
    }

    // =========================================================================
    // Promise with functional interface
    // =========================================================================

    /**
     * A promise for an async result.
     *
     * @param <T> the result type
     */
    public static final class Promise<T> {
        private final CompletableFuture<T> future;

        Promise(CompletableFuture<T> future) {
            this.future = future;
        }

        /**
         * Block until the result is available.
         */
        public T await() {
            try {
                return future.join();
            } catch (Exception e) {
                Throwable cause = e.getCause();
                if (cause instanceof DotDoException dde) {
                    throw dde;
                }
                throw new DotDoException(e.getMessage(), e);
            }
        }

        /**
         * Block with a timeout.
         */
        public T await(Duration timeout) {
            try {
                return future.get(timeout.toMillis(), TimeUnit.MILLISECONDS);
            } catch (java.util.concurrent.TimeoutException e) {
                throw new DotDoException("Request timed out after " + timeout);
            } catch (Exception e) {
                Throwable cause = e.getCause();
                if (cause instanceof DotDoException dde) {
                    throw dde;
                }
                throw new DotDoException(e.getMessage(), e);
            }
        }

        /**
         * Transform the result.
         */
        public <U> Promise<U> map(Function<T, U> mapper) {
            return new Promise<>(future.thenApply(mapper));
        }

        /**
         * Chain another async operation.
         */
        public <U> Promise<U> flatMap(Function<T, Promise<U>> mapper) {
            return new Promise<>(future.thenCompose(value -> mapper.apply(value).toFuture()));
        }

        /**
         * Handle errors.
         */
        public Promise<T> recover(Function<Throwable, T> handler) {
            return new Promise<>(future.exceptionally(handler));
        }

        /**
         * Get the underlying future.
         */
        public CompletableFuture<T> toFuture() {
            return future;
        }
    }

    // =========================================================================
    // Response
    // =========================================================================

    /**
     * API response.
     */
    public record Response(Object result, Object error) {
        /**
         * Get the result, throwing if there was an error.
         */
        public Object getResult() {
            if (error != null) {
                throw new DotDoException("API error: " + error);
            }
            return result;
        }

        /**
         * Check if successful.
         */
        public boolean isSuccess() {
            return error == null;
        }

        /**
         * Check if error.
         */
        public boolean isError() {
            return error != null;
        }
    }

    // =========================================================================
    // Authentication Manager
    // =========================================================================

    /**
     * Manages authentication tokens with automatic refresh.
     */
    static final class AuthManager {
        private final Config config;
        private final AtomicReference<TokenInfo> tokenRef = new AtomicReference<>();
        private final HttpClient httpClient = HttpClient.newHttpClient();

        AuthManager(Config config) {
            this.config = config;
        }

        String getToken() {
            // If API key is set, use it directly
            if (config.apiKey() != null) {
                return config.apiKey();
            }

            // Otherwise, use OAuth2 token
            TokenInfo token = tokenRef.get();
            if (token != null && !token.isExpired()) {
                return token.accessToken();
            }

            // Refresh token
            synchronized (this) {
                token = tokenRef.get();
                if (token == null || token.isExpired()) {
                    token = refreshToken();
                    tokenRef.set(token);
                }
            }
            return token.accessToken();
        }

        void invalidateToken() {
            tokenRef.set(null);
        }

        private TokenInfo refreshToken() {
            // OAuth2 client credentials flow
            String body = "grant_type=client_credentials" +
                    "&client_id=" + config.clientId() +
                    "&client_secret=" + config.clientSecret();

            var request = HttpRequest.newBuilder()
                    .uri(URI.create(config.baseUrl() + "/oauth/token"))
                    .header("Content-Type", "application/x-www-form-urlencoded")
                    .POST(HttpRequest.BodyPublishers.ofString(body))
                    .build();

            try {
                var response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
                if (response.statusCode() != 200) {
                    throw new AuthException("OAuth token refresh failed: " + response.body());
                }

                // Parse token response (simplified)
                String responseBody = response.body();
                String accessToken = extractStringValue(responseBody, "access_token");
                long expiresIn = extractLongValue(responseBody, "expires_in", 3600);

                return new TokenInfo(
                        accessToken,
                        Instant.now().plusSeconds(expiresIn - 60) // Refresh 60s before expiry
                );
            } catch (Exception e) {
                throw new AuthException("Failed to refresh token: " + e.getMessage(), e);
            }
        }

        private String extractStringValue(String json, String key) {
            int idx = json.indexOf("\"" + key + "\":");
            if (idx < 0) return null;
            int start = json.indexOf("\"", idx + key.length() + 3) + 1;
            int end = json.indexOf("\"", start);
            return json.substring(start, end);
        }

        private long extractLongValue(String json, String key, long defaultValue) {
            int idx = json.indexOf("\"" + key + "\":");
            if (idx < 0) return defaultValue;
            int start = idx + key.length() + 3;
            int end = start;
            while (end < json.length() && Character.isDigit(json.charAt(end))) {
                end++;
            }
            try {
                return Long.parseLong(json.substring(start, end));
            } catch (NumberFormatException e) {
                return defaultValue;
            }
        }

        record TokenInfo(String accessToken, Instant expiresAt) {
            boolean isExpired() {
                return Instant.now().isAfter(expiresAt);
            }
        }
    }

    // =========================================================================
    // Connection Pool
    // =========================================================================

    /**
     * Simple connection pool for managing HTTP connections.
     */
    static final class ConnectionPool implements AutoCloseable {
        private final int maxConnections;
        private final AtomicInteger activeConnections = new AtomicInteger(0);
        private final ConcurrentHashMap<Integer, ConnectionHandle> connections = new ConcurrentHashMap<>();
        private final AtomicInteger handleIdGenerator = new AtomicInteger(0);

        ConnectionPool(int maxConnections) {
            this.maxConnections = maxConnections;
        }

        ConnectionHandle acquire() {
            int current = activeConnections.get();
            if (current >= maxConnections) {
                // Wait for a connection to become available
                // In production, use a proper blocking queue
                while (activeConnections.get() >= maxConnections) {
                    try {
                        Thread.sleep(10);
                    } catch (InterruptedException e) {
                        Thread.currentThread().interrupt();
                        throw new DotDoException("Interrupted while waiting for connection");
                    }
                }
            }

            activeConnections.incrementAndGet();
            int id = handleIdGenerator.incrementAndGet();
            var handle = new ConnectionHandle(id);
            connections.put(id, handle);
            return handle;
        }

        void release(ConnectionHandle handle) {
            if (connections.remove(handle.id()) != null) {
                activeConnections.decrementAndGet();
            }
        }

        @Override
        public void close() {
            connections.clear();
            activeConnections.set(0);
        }

        record ConnectionHandle(int id) {}
    }

    // =========================================================================
    // Retry Policy
    // =========================================================================

    /**
     * Implements retry logic with exponential backoff.
     */
    static final class RetryPolicy {
        private final int maxRetries;
        private final Duration baseDelay;

        RetryPolicy(int maxRetries, Duration baseDelay) {
            this.maxRetries = maxRetries;
            this.baseDelay = baseDelay;
        }

        <T> CompletableFuture<T> execute(Supplier<CompletableFuture<T>> operation) {
            return executeWithRetry(operation, 0);
        }

        private <T> CompletableFuture<T> executeWithRetry(Supplier<CompletableFuture<T>> operation, int attempt) {
            return operation.get().exceptionallyCompose(error -> {
                if (attempt >= maxRetries) {
                    return CompletableFuture.failedFuture(error);
                }

                // Check if error is retryable
                if (!isRetryable(error)) {
                    return CompletableFuture.failedFuture(error);
                }

                // Calculate delay with exponential backoff
                long delayMs = baseDelay.toMillis() * (1L << attempt);

                return CompletableFuture.runAsync(() -> {
                    try {
                        Thread.sleep(delayMs);
                    } catch (InterruptedException e) {
                        Thread.currentThread().interrupt();
                    }
                }).thenCompose(ignored -> executeWithRetry(operation, attempt + 1));
            });
        }

        private boolean isRetryable(Throwable error) {
            // Retry on connection errors and 5xx server errors
            if (error instanceof DotDoException dde) {
                String message = dde.getMessage();
                return message != null && (
                        message.contains("HTTP 5") ||
                        message.contains("Connection") ||
                        message.contains("timeout")
                );
            }
            return error instanceof java.io.IOException ||
                   error instanceof java.net.ConnectException ||
                   error instanceof java.net.SocketTimeoutException;
        }
    }

    // =========================================================================
    // Client Wrappers
    // =========================================================================

    /**
     * Client wrapper with custom retry settings.
     */
    public static final class RetryClient {
        private final DotDo client;
        private final RetryPolicy retryPolicy;

        RetryClient(DotDo client, int maxRetries) {
            this.client = client;
            this.retryPolicy = new RetryPolicy(maxRetries, client.config.retryDelay());
        }

        public Promise<Response> call(String method) {
            return call(method, null);
        }

        public Promise<Response> call(String method, Object params) {
            client.ensureNotClosed();
            return new Promise<>(retryPolicy.execute(() -> client.call(method, params).toFuture()));
        }
    }

    /**
     * Client wrapper with custom timeout.
     */
    public static final class TimeoutClient {
        private final DotDo client;
        private final Duration timeout;

        TimeoutClient(DotDo client, Duration timeout) {
            this.client = client;
            this.timeout = timeout;
        }

        public Promise<Response> call(String method) {
            return call(method, null);
        }

        public Promise<Response> call(String method, Object params) {
            var promise = client.call(method, params);
            return new Promise<>(promise.toFuture().orTimeout(timeout.toMillis(), TimeUnit.MILLISECONDS));
        }
    }

    // =========================================================================
    // Exceptions
    // =========================================================================

    /**
     * Base exception for DotDo operations.
     */
    public static class DotDoException extends RuntimeException {
        public DotDoException(String message) {
            super(message);
        }

        public DotDoException(String message, Throwable cause) {
            super(message, cause);
        }
    }

    /**
     * Authentication exception.
     */
    public static class AuthException extends DotDoException {
        public AuthException(String message) {
            super(message);
        }

        public AuthException(String message, Throwable cause) {
            super(message, cause);
        }
    }
}

File: packages/kotlin/dotdo/src/main/kotlin/com/dotdo/DotDo.kt
package com.dotdo

import kotlinx.coroutines.*
import kotlinx.coroutines.channels.*
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.sync.*
import kotlinx.serialization.*
import kotlinx.serialization.json.*
import kotlin.time.Duration
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.seconds

/**
 * DotDo Platform SDK for Kotlin
 *
 * The main entry point for the DotDo platform, providing:
 * - Authentication and token management
 * - Connection pooling with automatic health checks
 * - Retry logic with exponential backoff
 * - Coroutine-based async operations
 */
object DotDo {

    private val defaultConfig = DotDoConfig()

    /**
     * Create a new DotDo client instance.
     */
    suspend fun connect(configure: DotDoConfig.() -> Unit = {}): DotDoClient {
        val config = DotDoConfig().apply(configure)
        return DotDoClient(config)
    }

    /**
     * Create a client with a specific endpoint.
     */
    suspend fun connect(endpoint: String, configure: DotDoConfig.() -> Unit = {}): DotDoClient {
        val config = DotDoConfig().apply {
            this.endpoint = endpoint
            configure()
        }
        return DotDoClient(config)
    }
}

/**
 * Configuration for DotDo client.
 */
class DotDoConfig {
    var endpoint: String = "https://api.dotdo.com"
    var authProvider: AuthProvider? = null
    var connectionPool: ConnectionPoolConfig = ConnectionPoolConfig()
    var retry: RetryConfig = RetryConfig()
    var timeout: Duration = 30.seconds

    /**
     * Configure authentication.
     */
    fun auth(configure: AuthConfig.() -> Unit) {
        val authConfig = AuthConfig().apply(configure)
        authProvider = when {
            authConfig.apiKey != null -> AuthProvider.ApiKey(authConfig.apiKey!!)
            authConfig.token != null -> AuthProvider.BearerToken(authConfig.token!!)
            authConfig.oauth != null -> authConfig.oauth!!
            else -> null
        }
    }

    /**
     * Configure connection pooling.
     */
    fun pool(configure: ConnectionPoolConfig.() -> Unit) {
        connectionPool = ConnectionPoolConfig().apply(configure)
    }

    /**
     * Configure retry behavior.
     */
    fun retry(configure: RetryConfig.() -> Unit) {
        retry = RetryConfig().apply(configure)
    }
}

/**
 * Authentication configuration.
 */
class AuthConfig {
    var apiKey: String? = null
    var token: String? = null
    var oauth: AuthProvider.OAuth? = null

    fun apiKey(key: String) {
        apiKey = key
    }

    fun bearerToken(token: String) {
        this.token = token
    }

    fun oauth(configure: OAuthConfig.() -> Unit) {
        val config = OAuthConfig().apply(configure)
        oauth = AuthProvider.OAuth(
            clientId = config.clientId,
            clientSecret = config.clientSecret,
            tokenUrl = config.tokenUrl,
            scopes = config.scopes
        )
    }
}

class OAuthConfig {
    var clientId: String = ""
    var clientSecret: String = ""
    var tokenUrl: String = ""
    var scopes: List<String> = emptyList()
}

/**
 * Authentication providers.
 */
sealed class AuthProvider {
    abstract suspend fun getAuthHeader(): String

    data class ApiKey(val key: String) : AuthProvider() {
        override suspend fun getAuthHeader(): String = "X-API-Key $key"
    }

    data class BearerToken(val token: String) : AuthProvider() {
        override suspend fun getAuthHeader(): String = "Bearer $token"
    }

    data class OAuth(
        val clientId: String,
        val clientSecret: String,
        val tokenUrl: String,
        val scopes: List<String>
    ) : AuthProvider() {
        private var cachedToken: String? = null
        private var tokenExpiry: Long = 0

        override suspend fun getAuthHeader(): String {
            if (cachedToken == null || System.currentTimeMillis() >= tokenExpiry) {
                refreshToken()
            }
            return "Bearer $cachedToken"
        }

        private suspend fun refreshToken() {
            // TODO: Implement OAuth token refresh
            throw NotImplementedError("OAuth token refresh not yet implemented")
        }
    }
}

/**
 * Connection pool configuration.
 */
class ConnectionPoolConfig {
    var maxConnections: Int = 10
    var maxIdleTime: Duration = 5.seconds
    var healthCheckInterval: Duration = 30.seconds
    var acquireTimeout: Duration = 10.seconds
}

/**
 * Retry configuration with exponential backoff.
 */
class RetryConfig {
    var maxAttempts: Int = 3
    var initialDelay: Duration = 100.milliseconds
    var maxDelay: Duration = 10.seconds
    var multiplier: Double = 2.0
    var retryOn: Set<Class<out Throwable>> = setOf(
        java.io.IOException::class.java,
        java.net.SocketTimeoutException::class.java
    )

    fun retryOn(vararg exceptions: Class<out Throwable>) {
        retryOn = exceptions.toSet()
    }
}

/**
 * Main DotDo client.
 */
class DotDoClient internal constructor(
    private val config: DotDoConfig
) : AutoCloseable {

    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private val connectionPool = ConnectionPool(config.connectionPool)
    private val json = Json { ignoreUnknownKeys = true }
    private var closed = false

    init {
        // Start connection pool health checks
        scope.launch {
            connectionPool.startHealthChecks(config.connectionPool.healthCheckInterval)
        }
    }

    /**
     * Execute an operation with automatic retry.
     */
    suspend fun <T> withRetry(block: suspend () -> T): T {
        return retryWithBackoff(config.retry, block)
    }

    /**
     * Get an authenticated connection from the pool.
     */
    suspend fun <T> withConnection(block: suspend (Connection) -> T): T {
        val connection = connectionPool.acquire(config.connectionPool.acquireTimeout)
        return try {
            // Add auth header if configured
            config.authProvider?.let { auth ->
                connection.headers["Authorization"] = auth.getAuthHeader()
            }
            block(connection)
        } finally {
            connectionPool.release(connection)
        }
    }

    /**
     * Check if the client is connected and healthy.
     */
    fun isHealthy(): Boolean = !closed && connectionPool.hasHealthyConnections()

    /**
     * Get connection pool statistics.
     */
    fun poolStats(): PoolStats = connectionPool.stats()

    override fun close() {
        closed = true
        scope.cancel()
        connectionPool.close()
    }

    // Retry with exponential backoff
    private suspend fun <T> retryWithBackoff(
        config: RetryConfig,
        block: suspend () -> T
    ): T {
        var currentDelay = config.initialDelay
        var lastException: Throwable? = null

        repeat(config.maxAttempts) { attempt ->
            try {
                return block()
            } catch (e: Throwable) {
                lastException = e
                val shouldRetry = config.retryOn.any { it.isInstance(e) }
                if (!shouldRetry || attempt == config.maxAttempts - 1) {
                    throw e
                }

                delay(currentDelay)
                currentDelay = minOf(
                    currentDelay * config.multiplier.toLong(),
                    config.maxDelay
                )
            }
        }

        throw lastException ?: IllegalStateException("Retry failed with no exception")
    }
}

/**
 * A pooled connection.
 */
class Connection internal constructor(
    internal val id: String,
    internal val createdAt: Long = System.currentTimeMillis()
) {
    internal var lastUsed: Long = System.currentTimeMillis()
    internal var healthy: Boolean = true
    internal val headers: MutableMap<String, String> = mutableMapOf()

    fun isHealthy(): Boolean = healthy

    internal fun markUsed() {
        lastUsed = System.currentTimeMillis()
    }
}

/**
 * Connection pool with health checks.
 */
class ConnectionPool internal constructor(
    private val config: ConnectionPoolConfig
) {
    private val connections = Channel<Connection>(config.maxConnections)
    private val allConnections = mutableListOf<Connection>()
    private val mutex = Mutex()
    private var connectionCounter = 0

    init {
        // Pre-populate pool
        repeat(config.maxConnections) {
            val conn = createConnection()
            allConnections.add(conn)
            connections.trySend(conn)
        }
    }

    suspend fun acquire(timeout: Duration): Connection {
        return withTimeout(timeout) {
            val conn = connections.receive()
            conn.markUsed()
            conn
        }
    }

    fun release(connection: Connection) {
        connection.headers.clear()
        connections.trySend(connection)
    }

    fun hasHealthyConnections(): Boolean {
        return allConnections.any { it.healthy }
    }

    fun stats(): PoolStats {
        return PoolStats(
            totalConnections = allConnections.size,
            healthyConnections = allConnections.count { it.healthy },
            availableConnections = connections.isEmpty.let { if (it) 0 else allConnections.size }
        )
    }

    suspend fun startHealthChecks(interval: Duration) {
        while (true) {
            delay(interval)
            checkHealth()
        }
    }

    private suspend fun checkHealth() {
        mutex.withLock {
            val now = System.currentTimeMillis()
            allConnections.forEach { conn ->
                // Mark as unhealthy if idle too long
                val idleTime = now - conn.lastUsed
                if (idleTime > config.maxIdleTime.inWholeMilliseconds) {
                    conn.healthy = false
                }
            }

            // Replace unhealthy connections
            val unhealthy = allConnections.filter { !it.healthy }
            unhealthy.forEach { old ->
                allConnections.remove(old)
                val newConn = createConnection()
                allConnections.add(newConn)
                connections.trySend(newConn)
            }
        }
    }

    private fun createConnection(): Connection {
        return Connection(id = "conn-${++connectionCounter}")
    }

    fun close() {
        connections.close()
        allConnections.clear()
    }
}

/**
 * Pool statistics.
 */
data class PoolStats(
    val totalConnections: Int,
    val healthyConnections: Int,
    val availableConnections: Int
)

/**
 * DotDo-specific errors.
 */
sealed class DotDoException : Exception() {
    data class AuthenticationFailed(val reason: String) : DotDoException() {
        override val message: String = "Authentication failed: $reason"
    }

    data class ConnectionPoolExhausted(val timeout: Duration) : DotDoException() {
        override val message: String = "Connection pool exhausted after $timeout"
    }

    data class ServiceUnavailable(val service: String) : DotDoException() {
        override val message: String = "Service unavailable: $service"
    }

    data class RateLimited(val retryAfter: Duration?) : DotDoException() {
        override val message: String = "Rate limited${retryAfter?.let { ", retry after $it" } ?: ""}"
    }
}

/**
 * Extension function to run operations with a fresh client.
 */
suspend fun <T> withDotDo(
    configure: DotDoConfig.() -> Unit = {},
    block: suspend DotDoClient.() -> T
): T {
    return DotDo.connect(configure).use { client ->
        block(client)
    }
}
